notebooks/working/3.0-mab-independent-variable-simulation.md

```diff
@@ -34,487 +47,269 @@ import random
-#         Name of the column containing the observation
-#         id numbers in the long format dataset.
-#     number_alts: int
-#         Number of alternatives in the long format
-#         dataset.
-#     spec_dic: dictionary
-#         Dictionary of the model specification.
-#     names_dic: dictionary
-#         Dictionary of the alternative names.
-#     init_betas: list
-#         List of the initial betas for the model
-#         from which the choices should be simulated.
+## Functions to replace code within
+## FindLongDataDist
+## Function for checking length
+def isUnique(var_values):
+    return len(var_values.unique()) == 1
+
+## Function for checking if variable is categorical
+def isCategorical(var_type):
+    return var_type == 'categorical'
+
+## Function to get distribution for unique values:
+def getConstantDist(ind_var, var_val):
+    constant_dict = defaultdict(dict)
+    constant_dict[ind_var]['distribution'] = 'constant'
+    constant_dict[ind_var]['parameters'] = var_val.unique()
+    return constant_dict
+
+## Function to get distribution for categorical variables:
+def getCategoricalDist(ind_var, var_val):
+    categ_dict = defaultdict(dict)
+    categ_dict[ind_var]['distribution'] = 'categorical'
+    # Count frequency of values and store it as paramater of distribution
+    np_array_range = np.arange(var_val.max()+1)
+    array_bincount = np.bincount(var_val)
+    probs = array_bincount / len(var_val)
+    categ_dict[ind_var]['parameters'] = [np_array_range,
+                                      probs]
+    return categ_dict
+
+
+def getContinuousDist(ind_var, var_val, cont_dists):
+    cont_dict = defaultdict(dict)    
+    fitter_object = Fitter(data=var_val,
+                           distributions=cont_dists,
+                           timeout=30)
+    fitter_object.fit()
+    # Get the best distribution and store in dictionary
+    BestDict = fitter_object.get_best()
+    cont_dict[ind_var]['distribution'] = list(BestDict.items())[0][0]
+    cont_dict[ind_var]['parameters'] = list(BestDict.items())[0][1]
+    return cont_dict
+
+## Function for looping around individual specific variables
+def indSpecDist(data_long, obs_id_col, ind_spec, var_types):
+    ind_spec_dict = defaultdict(dict)
+    for ind_var in ind_spec:
+        # generate array of values for individual specific variable
+        var_val = pd.Series([(data_long.loc[data_long[obs_id_col] == x][ind_var].unique()[0]) for x in data_long[obs_id_col].unique()])
+        var_type = var_types[ind_var]
+        # Get distribution if variable is categorical
+        if isCategorical(var_type):
+            # If only one category
+            if isUnique(var_val):
+                ind_var_dict = getConstantDist(ind_var, var_val)
+            # If more than one category
+            else:
+                ind_var_dict = getCategoricalDist(ind_var, var_val)
+        else:
+            ind_var_dict = getContinuousDist(ind_var, var_val, cont_dists)
```

Comment:
I believe you are relying on a global variable here (which you shouldn't do).
`cont_dists` is not defined as one of the function inputs, and it is not generated between lines 100 and 114. Try adding `cont_dists` to your inputs!

Reply:
Great catch!